File: two_phase_batch_model/two_phase_batch_model.cpp
#include<iostream>
#include<vector>
#include<fstream>


// parameters that are "fixed"
const double MwE = 65000.0;
const double MwG = 162.0;
const double Mwg = 180.0;
const double rhol = 1000.0;
const double rhoT = rhol;
const double rGg = MwG/Mwg;

// kinetics model parameters
const double kF = 5000.0;
const double kR = 1.0*kF;
const double KdF = 0.03773;
const double KdR = 0.5079;
const double KI = 0.01;


void get_adsorption(std::vector<double>& adsorption, double CGF, double CGRA, double CET, double cg, double epsl) {
    /*
    Adsorption equilibrium of enzyme between facile and recalcitrant glucan,
    and accounting for inhibition by glucose (and other sugars if present)
    */
    
    double CEGF = CET/(1.0 + KdF/KdR*CGRA/CGF + epsl*KdF/CGF*(1.0 + cg/KI));
    double CEGR = CET/(1.0 + KdR/KdF*CGF/CGRA + epsl*KdR/CGRA*(1.0 + cg/KI));
    
    adsorption[0] = CEGF;
    adsorption[1] = CEGR;
}


void get_rhs(std::vector<double>& rhs, std::vector<double> solnvec, double fET, double fX0, double fL0) {
    /*
    RHS of system of ODES
    */

    // unpack
    double fGF = solnvec[0];
    double fGR = solnvec[1];
    double fg = solnvec[2];
    
    // convenience term
    double mwGrho = MwG/rhoT;
    
    // molar concentrations
    double CGF = fGF/mwGrho;
    double CGR = fGR/mwGrho;
    double CET = rhoT/MwE*fET;
    
    // calculate liquid molar sugar concentration
    double fG = fGF + fGR;
    double fis = fG + fX0 + fL0;
    double fliq = 1.0 - fis;
    double epsl = rhoT/rhol*fliq;
    double cg = rhol/fliq/Mwg*fg;

    // enzyme adsorption -- with CGRA = CGR (no limited exposure)
    std::vector<double> adsorption(2);
    get_adsorption(adsorption, CGF, CGR, CET, cg, epsl);
    double rF = kF*adsorption[0];
    double rR = kR*adsorption[1];

    // ODEs
    double dfGF = -mwGrho*rF;
    double dfGR = -mwGrho*rR;
    double dfg = Mwg/rhoT*(rF + rR);
    
    rhs[0] = dfGF;
    rhs[1] = dfGR;
    rhs[2] = dfg;
}
    

void advance_soln(std::vector<double>& solnvec, double fET, double fX0, double fL0, double dt, int nvars) {

    std::vector<double> solnvec_0(nvars);
    std::vector<double> k1(nvars);
    std::vector<double> k2(nvars);
    std::vector<double> k3(nvars);
    std::vector<double> k4(nvars);

    // Store the initial value
    solnvec_0 = solnvec;

    // Calculate k1: the slope at the beginning of the interval
    get_rhs(k1, solnvec_0, fET, fX0, fL0);

    // Calculate k2: the slope at the midpoint of the interval following k1    
    for(int i=0; i<nvars; i++) {
        solnvec[i] = solnvec_0[i] + 0.5*dt*k1[i];            
    }

    get_rhs(k2, solnvec, fET, fX0, fL0);

    // Calculate k3: the slope at the midpoint of the interval following k2
    for(int i=0; i<nvars; i++) {
        solnvec[i] = solnvec_0[i] + 0.5*dt*k2[i];            
    }

    get_rhs(k3, solnvec, fET, fX0, fL0);

    // Calculate k4: the slope at the end of the interval following k3
    for(int i=0; i<nvars; i++) {
        solnvec[i] = solnvec_0[i] + dt*k3[i];            
    }

    get_rhs(k4, solnvec, fET, fX0, fL0);

    // Update the solution
    for(int i=0; i<nvars; i++) {
        solnvec[i] = solnvec_0[i] + 1.0/6.0*dt*(k1[i] + 2.0*k2[i] + 2.0*k3[i] + k4[i]);          
    }
}


int main() {

    // Declare time-stepping parameters
    double t_final = 100.0;
    double dt = 0.1;
    double tt = 0.0;
    int t_steps = (int)(t_final/dt);

    // These values should be obtained from the notebook
    double lmbde = 0.03;
    double fis0 = 0.05;
    double dilution_factor = 0.5;
    double xG0 = 1.0;
    double xX0 = 0.0;
    double rhog0 = 0.0;
    double rhox0 = 1.0*dilution_factor;
    double rhof0 = 0.0*dilution_factor;

    double conversion_xylan = 0.5;
    double yF0 = 0.2 + 0.6*conversion_xylan;

    // initial conditions in model variables
    double fG0 = xG0*fis0;
    double fGF0 = yF0*fG0;
    double fGR0 = (1-yF0)*fG0;
    double fX0 = xX0*fis0;
    double fL0 = (1 - xG0 - xX0)*fis0;
    double fliq0 = 1 - fis0;
    double fg0 = rhog0*fliq0/rhol;
    double fET = lmbde*fG0;

    // Initialize all vectors
    int nvars = 3;
    std::vector<double> solnvec(nvars);

    // Store initial conditions
    solnvec[0] = fGF0;
    solnvec[1] = fGR0;
    solnvec[2] = fg0;

    // Initialize output file
    std::ofstream outfile("eh_results.csv");
    outfile.precision(9);

    // Write header line
    outfile<<"# time, fGF, fGR, fg, rhog\n";  

    // Save initial condition
    outfile<<0.0<<", ";
    for(int i=0; i<nvars; i++) {
        outfile<<solnvec[i]<<", ";
    }
    outfile<<0.0<<"\n";


    double fGF, fGR, fg;
    double fG, fis, fliq, rhog; 
    double convF, convR, conv, mG0, mG, mb;

    for(int i=0; i<t_steps; i++) {
        advance_soln(solnvec, fET, fX0, fL0, dt, nvars);
        tt += dt;

        // unpack and convert output quantities (maybe not all necessary in solver)
        fGF = solnvec[0];
        fGR = solnvec[1];
        fg = solnvec[2];

        fG = fGF + fGR;
        fis = fG + fX0 + fL0;
        fliq = 1.0 - fis;
        rhog = rhol/fliq*fg;

        convF = (fGF0 - fGF)/fG0;
        convR = (fGR0 - fGR)/fG0;
        conv = 1.0 - fG/fG0;
        mG0 = fG0 + fg0*rGg;
        mG = fG + fg*rGg;
        mb = 1.0 - mG/mG0;

        // Write data for this timestep
        outfile<<tt<<", ";
        for(int i=0; i<nvars; i++) {
            outfile<<solnvec[i]<<", ";
        }
        outfile<<rhog<<"\n";
    }


    outfile.close();

    return 0;
}

File: applications/Bioconversion/models/two_phase_batch_model/two_phase_batch_model.py
"""
Test two-phase reaction rate model for EH. This script applies it to a batch
reaction

"""

# Jonathan Stickel, 2017


#TODO:  put the model in a module!


# variable convention:
# Ci = c^tilde = mol i per total slurry volume (mol/L)
# ci = mol i per liquid volume (mol/L)
# rhoi = mass i per liquid volume (g/L)
# fi = mass i per total slurry mass (g/g)


import numpy as np
from scipy import integrate as igt
from scipy import optimize as opt

from matplotlib.pyplot import *
import matplotlib as mpl

import sys
from vebio.Utilities import dict_to_yaml, yaml_to_dict


if len(sys.argv) > 1:
    params_filename = sys.argv[1]
    ve_params = yaml_to_dict(params_filename)

else:
    ve_params = {} # won't this just cause errors? JJS 3/14/21


font={'family':'Helvetica', 'size':'15'}
mpl.rc('font',**font)
mpl.rc('xtick',labelsize=14)
mpl.rc('ytick',labelsize=14)


# parameters that are "fixed"
MwE = 65e3 # mol weight enzymes (CBH, but close to EG and BG)
# Jeoh, T., Michener, W., Himmel, M. E., Decker, S. R., & Adney, W. S. (2008). Implications of cellobiohydrolase glycosylation for use in biomass conversion. Biotechnology for Biofuels, 1, 10. https://doi.org/10.1186/1754-6834-1-10
MwG = 162. # mol weight glucan
Mwg = 180. # mol weight glucose
rhol = 1000. # kg/m^3 = g/L
rhoT = rhol # reasonable approximation -- most terms that need rhoT cancel
            # errors, so that this approximation provides near machine
            # precision accuracy
rGg = MwG/Mwg

# kinetics model parameters
#kF = 1000. # 1/h, trial-and-error
kF = 5e3 # best fit (constrained)
kR = 1e0*kF # 1/h
## Kdi and KI are "desorption" equilibrium constants -- lower values means more
## adsorbed enzyme
#KdF = 0.05 # mol/L = kmol/m^3 , trial-and-error
KdF = 0.03773 # kmol/m^3, best fit
#KdR = 10*KdF 
KdR = 0.5079 # kmol/m^3, best fit
#KI = 1e6 # kmol/m^3; high value means less inhibition -- 1e6 = no inhibition
#KI = 0.05 # kmol/m^3; trial-and-error
KI = 1e-2 # best fit (constrained)
# lignin solubilization parameters -- not used yet
#kL = 0.6

# reactor starting conditions
#mT0 = 10. # kg -- not used


#lmbde = 0.03 # kg/kg; *1000 to get mg/g
# Direct User Input
lmbde = ve_params['enzymatic_input']['lambda_e']

#fis0 = 0.05 # kg/kg; initial fraction insoluble solids
# Direct User Input (note, this is a target, not the output from pretreatment)
fis0 = ve_params['enzymatic_input']['fis_0']

# Compute the amount of dilution required to reach the fis_0_target
# based on the output from the pretreatment step
dilution_factor = fis0/ve_params['pretreatment_output']['fis_0']

#xG0 = 1.0 # initial glucan fraction of insoluble solids -- 100% here
xG0 = ve_params['pretreatment_output']['X_G']
xX0 = ve_params['pretreatment_output']['X_X']
rhog0 = 0.0 # g/L; initial glucose concentration in the liquid
#rhox0 = 0.0 # g/L
rhox0 = ve_params['pretreatment_output']['rho_x']*dilution_factor
rhof0 = ve_params['pretreatment_output']['rho_f']*dilution_factor # furfural

#yF0 = 0.4 # fraction of glucan that is "facile" -- best fit (constrained)
conversion_xylan = ve_params['pretreatment_output']['conv']
yF0 = 0.2 + 0.6*conversion_xylan
# yF0 = 1.0*conversion_xylan

print('\nINPUTS')
print('Lambda_e = %.4f' % (lmbde))
print('FIS_0 = %.4f' % (fis0))
print('yF0 = %.4f' % (yF0))


# initial conditions in model variables
fG0 = xG0*fis0
fGF0 = yF0*fG0
fGR0 = (1-yF0)*fG0
fX0 = xX0*fis0 # used only for fis calculations
fL0 = (1 - xG0 - xX0)*fis0 # ditto
fliq0 = 1 - fis0
#epsl0 = rhoT/rhol*fliq0 # not needed
fg0 = rhog0*fliq0/rhol
#fx0 = rhox0*fliq0/rhoT
#fsl0 = 0
fET = lmbde*fG0
f0 = np.array([fGF0, fGR0, fg0])


# enzymatic hydrolysis rate equation
def adsorption(CGF, CGRA, CET, cg, epsl, KdF, KdR, KI):
    """
    Adsorption equilibrium of enzyme between facile and recalcitrant glucan,
    and accounting for inhibition by glucose (and other sugars if present)

    """
    CEGF = CET/(1 + KdF/KdR*CGRA/CGF + epsl*KdF/CGF*(1 + cg/KI))
    CEGR = CET/(1 + KdR/KdF*CGF/CGRA + epsl*KdR/CGRA*(1 + cg/KI))
    return CEGF, CEGR


def rhs(f, t, rhoT, fET, kF, kR, KdF, KdR, KI):
    """ RHS of system of ODES"""
    # unpack
    fGF, fGR, fg = f

    # convenience term
    mwGrho = MwG/rhoT
    
    # molar concentrations
    CGF = fGF/mwGrho # rhoT/MwG*fGF
    CGR = fGR/mwGrho # rhoT/MwG*fGR
    CET = rhoT/MwE*fET
    
    # calculate liquid molar sugar concentration
    fG = fGF + fGR
    fis = fG + fX0 + fL0
    fliq = 1 - fis
    epsl = rhoT/rhol*fliq 
    cg = rhol/fliq/Mwg*fg

    # enzyme adsorption -- with CGRA = CGR (no limited exposure)
    CEGF, CEGR = adsorption(CGF, CGR, CET, cg, epsl, KdF, KdR, KI)
    rF = kF*CEGF
    rR = kR*CEGR

    # ODEs
    dfGF = -mwGrho*rF
    dfGR = -mwGrho*rR
    dfg = Mwg/rhoT*(rF + rR)
    
    return np.array([dfGF, dfGR, dfg])

# deleted model fitting -- would not do that with virtual engineering sims, JJS 3/23/20


# run simulation via igt.odeint
#tfin = 200. # h
# tfin = 100. # h
tfin = ve_params['enzymatic_input']['t_final']
print('t_final = %.4f' % (tfin))

N = 200
t = np.linspace(0, tfin, N)
f = igt.odeint(rhs, f0, t, args=(rhoT, fET, kF, kR, KdF, KdR, KI))

fGF = f[:,0]
fGR = f[:,1]
fg = f[:,2]

fG = fGF + fGR
fis = fG + fX0 + fL0
fliq = 1 - fis
#epsl = rhoT/rhol*fliq 
rhog = rhol/fliq*fg 

convF = (fGF0 - fGF)/fG0
convR = (fGR0 - fGR)/fG0
conv = 1 - fG/fG0
mG0 = fG0 + fg0*rGg
mG = fG + fg*rGg
mb = 1 - mG/mG0


# Save the outputs into a dictionary for use as inputs for bioreactor sims
output_dict = {'enzymatic_output': {}}
output_dict['enzymatic_output']['rho_g'] = float(rhog[-1])

# fixed dilution_EH factor, JJS 3/14/21
dilution_EH = fliq[0]/fliq[-1]
output_dict['enzymatic_output']['rho_x'] = float(rhox0*dilution_EH)
output_dict['enzymatic_output']['rho_f'] = float(rhof0*dilution_EH)

if len(sys.argv) > 1:
    dict_to_yaml([ve_params, output_dict], params_filename)

print('\nFINAL OUTPUTS (at t = %.1f hours)' % (tfin))
print('rho_g = %.4f' % (rhog[-1]))
# print('rho_x = ', rhox0*dilution_EH)
# print('rho_f = ', rhof0*dilution_EH)
# print('Facile Conversion = ', convF[-1])
# print('Recalcitrant Conversion = ', convR[-1])
print('Total Conversion = %.4f' % (conv[-1]))
print()


# save data to compare with other modeling approaches -- probably not needed for VE use
if False:
    #np.savez('two-phase_no_struc.npz', t=t, conv=conv, rhog=rhog)
    filename='two-phase_no_struc_'+str(fis0)+'.npz'
    np.savez(filename, t=t, conv=conv, rhog=rhog, fis0=fis0)
    
    outfile=open("wellmix_sim.dat","w")
    for i in range(len(conv)):
        outfile.write("%e\t%e\n"%(t[i],conv[i]))
    outfile.close()

if ve_params['enzymatic_input']['show_plots']:
    figure(1)
    clf()
    xlim((-5,50))
    ylim((-0.02,1))
    plot(t, conv, lw=3,label='total')
    #plot(t, convF, lw=3,marker='*',markevery=10,markersize=10,label='facile')
    plot(t, convF, '--', lw=3, label='facile')
    #plot(t, convR, lw=3,marker='^',markevery=10,markersize=10,label='recalcitrant')
    plot(t, convR, '-.', lw=3, label='recalcitrant')
#    plot(tdat, convdat, 'o', markersize=10,label='exp. data')
    xlabel('time (hours)')
    ylabel('conversion')
    legend(loc='best')
    #savefig("two-phase_conversion.pdf", bbox_inches='tight')
    

    figure(2)
    clf()
    plot(t, mb)
    xlabel('t [h]')
    ylabel('mass balance')

    figure(3)
    clf()
    plot(t, rhog, lw=2, label='model')
#    plot(tdat, rhosdat, 'o', label='data')
    xlabel('t [h]')
    ylabel(r'$\rho_g$ [g/L]')
    legend(loc='best')
    
    figure(4)
    clf()
    xlim((-5,50))
    ylim((-0.005,0.07))
    #plot(t, fGF+fGR, lw=3, marker='*',markevery=10,markersize=10,label='total glucan')
    #plot(t, fGF, lw=3,marker='s',markevery=10,markersize=10,label='facile')
    #plot(t, fGR, lw=3,marker='^',markevery=10,markersize=10,label='recalcitrant')
    plot(t, fGF+fGR, '--', lw=3, label='total glucan')
    plot(t, fGF, '-.', lw=3, label='facile')
    plot(t, fGR, lw=3, ls=(0,(6,2)), label='recalcitrant')
    #plot(t, fGR, lw=1, ls=(5,(5,2)), label='recalcitrant')
    plot(t, fg, lw=3, label='glucose')
    xlabel('time (hours)')
    ylabel('mass fraction')
    legend(loc='best')
    #savefig("two-phase_species.pdf", bbox_inches='tight')
    
    # figure(5)
    # clf()
    # plot(t, fGF/fGR, lw=2, label='model')
    # xlabel('t [h]')
    # #ylabel('?')
    # legend(loc='best')

    show()
